src/backend/cpu/TNJ/BinaryNode.hpp

```diff
@@ -44,14 +44,24 @@ namespace TNJ
         {
         }
 
-        void calc(int x, int y, int z, int w)
+        void calc(int x, int y, int z, int w, int lim)
         {
-            this->m_val = m_op.eval(m_lhs->m_val, m_rhs->m_val);
+            Ti *lhs_ptr = &(m_lhs->m_val.front());
+            Ti *rhs_ptr = &(m_rhs->m_val.front());
+            To *out_ptr = &(this->m_val.front());
+            for (int i = 0; i < lim; i++) {
+                out_ptr[i] = m_op.eval(lhs_ptr[i], rhs_ptr[i]);
+            }
         }
 
-        void calc(int idx)
+        void calc(int idx, int lim)
         {
-            this->m_val = m_op.eval(m_lhs->m_val, m_rhs->m_val);
+            Ti *lhs_ptr = &(m_lhs->m_val.front());
+            Ti *rhs_ptr = &(m_rhs->m_val.front());
+            To *out_ptr = &(this->m_val.front());
+            for (int i = 0; i < lim; i++) {
+                out_ptr[i] = m_op.eval(lhs_ptr[i], rhs_ptr[i]);
+            }
```

Comment:
One more thing I noticed is that our TNJ -  BinaryOp, UnaryOp (I understand it is consistent with CUDA & OpenCL) structures have  `std::function` variables (which replaced the `eval` I believe). Why don't we just make these Functors, that would transform the above lines of code into a single line(more or less) such as
```cpp
std::transform(m_lhs->m_val.begin(), m_lhs->m_val.end(),
               m_rhs->begin(), this->m_val.begin(), m_op);
```
Hopefully, this may reduce code in other places as well.